alibi/utils/distributed.py

```diff
@@ -217,3 +232,523 @@ def _chunk(values, chunksize):
+        self.pool = self.create_parallel_pool(
+            explainer_type,
+            explainer_init_args,
+            explainer_init_kwargs
+        )
+        # use to retrieve state from different actors
+        self._actor_index = 0
+
+    def __getattr__(self, item: str) -> Any:
+        """
+        Accesses actor attributes. Use sparingly as this involves a remote call (that is, these attributes are of an
+        object in a different process). *The intended use is for retrieving any common state across the actor at the end
+        of the computation in order to form the response (see notes 2 & 3)*.
+
+        Parameters
+        ----------
+        item
+            The explainer attribute to be returned.
+
+        Returns
+        -------
+            The value of the attribute specified by `item`.
+
+        Raises
+        ------
+        ValueError
+            If the actor index is invalid.
+
+        Notes
+        -----
+            1. This method assumes that the actor implements a `return_attribute` method.
+            2. Note that we are indexing the idle actors. This means that if a pool was initialised with 5 actors and 3
+            are busy, indexing with index 2 will raise an IndexError.
+            3. The order of _idle_actors constantly changes - an actor is removed from it if there is a task to execute
+            and appended back when the task is complete. Therefore, indexing at the same position as computation
+            proceeds will result in retrieving state from different processes.
+        """
+
+        if self._actor_index > self.n_processes - 1:
+            raise ValueError(f"Index of actor should be less than or equal to {self.n_processes - 1}!")
+
+        actor = self.pool._idle_actors[self._actor_index]  # noqa
+        return DistributedExplainer.ray.get(actor.return_attribute.remote(item))
+
+    @property
+    def actor_index(self) -> int:
+        """
+        Returns the index of the actor for which state is returned.
+        """
+        return self._actor_index
+
+    @actor_index.setter
+    def actor_index(self, value: int):
+        """
+        Sets the actor index to allow retrieving state from a specific actor.
+        """
+        self._actor_index = value
+
+    def set_actor_index(self, value: int):
+        """
+        Sets actor index. This is used when the DistributedExplainer is in a separate process because `ray` does not
+        support calling property setters remotely
+        """
+        self._actor_index = value
+
+    def return_attribute(self, name: str) -> Any:
+        """
+        Returns an attribute specified by its name. Used in a distributed context where the properties cannot be
+        accessed using the dot syntax.
+        """
+
+        # if the attribute requested if of the PoolCollection object, return it
+        try:
+            return self.__getattribute__(name)
```

Comment:
I'm not sure I follow, how is `PoolCollection` related to this?

Reply:
Sorry, this is garbled. You can do `my_pool_collection.prop` where `prop` is a either an attribute of the `PoolCollection` or of the explainer wrapped by the `DistributedExplainer`. `__geattr__` handles the latter case.